cnd/src/spawn.rs

```diff
@@ -69,58 +70,60 @@ pub async fn spawn(
-                    tracing::error!(
-                        swap = %swap_id,
-                        "failed to complete swap: {:#}",
-                        e
-                    );
-
-                    if let Err(e) = storage
-                        .db
-                        .do_in_transaction(|conn| {
-                            commands::update_order_of_swap_to_failed(conn, swap_id)
-                        })
-                        .await
-                    {
-                        tracing::error!("failed to update order state: {:#}", e);
-                    }
-                }
-                (Ok(Ok(())), Ok(Ok(()))) => {
-                    tracing::info!(
-                        swap = %swap_id,
-                        "swap completed"
-                    );
-
-                    if let Err(e) = storage
-                        .db
-                        .do_in_transaction(|conn| {
-                            commands::update_order_of_swap_to_closed(conn, swap_id)
-                        })
-                        .await
-                    {
-                        tracing::error!("failed to update order state: {:#}", e);
-                    }
-                }
-                (Err(e), _) | (_, Err(e)) => {
-                    let e = Box::new(e);
-                    tracing::error!(
-                        error = &e as &(dyn std::error::Error + 'static),
-                        swap = %swap_id,
-                        "runtime error while executing protocol futures",
-                    );
-                }
-            };
-        });
+        handle.spawn(swap_result_handler(
+            alpha_handle,
+            beta_handle,
+            storage,
+            swap_id,
+        ));
     });
 
     Ok(())
 }
 
+async fn swap_result_handler(
+    alpha: JoinHandle<Result<()>>,
+    beta: JoinHandle<Result<()>>,
+    storage: Storage,
+    swap_id: LocalSwapId,
+) {
+    let db_update: Box<dyn Fn(&SqliteConnection) -> Result<()> + Send> =
+        match future::try_join(alpha, beta).await {
+            // Join successful and both protocols finish successfully
+            Ok((Ok(()), Ok(()))) => {
+                tracing::info!(swap = %swap_id, "swap completed");
+
+                Box::new(move |conn| {
+                    commands::update_order_of_swap_to_closed(conn, swap_id)?;
+                    commands::mark_swap_as_completed(conn, swap_id, OffsetDateTime::now_utc())?;
+
+                    Ok(())
+                })
+            }
+            // Join successful but one of the protocols failed
+            Ok((Err(e), _)) | Ok((_, Err(e))) => {
+                tracing::error!(swap = %swap_id, "failed to complete swap: {:#}", e);
+
+                Box::new(move |conn| {
+                    commands::update_order_of_swap_to_failed(conn, swap_id)?;
+                    // we don't mark a swap as completed in case of failure so that a
```

Comment:
If I am not mistaken this behavior might cause trouble when trying to close an order if retrying succeeds? 

`update_order_of_swap_to_closed` is not capable of dealing with `failed`:

https://github.com/comit-network/comit-rs/blob/only-active-swaps/cnd/src/storage/db/commands.rs#L66-L91

and `update_btc_dai_order_to_settling` does not seem to deal with `failed` neither:

https://github.com/comit-network/comit-rs/blob/only-active-swaps/cnd/src/storage/db/commands.rs#L26-L42

Reply:
Yes true. The proper way of doing this would be to not keep track of these individual states but store the whole swap execution in the DB and reason about it there.

As a patch, I could change the order state again once the swap is re-spawned?
Do we care much at this stage? :sweat_smile: